### PR TITLE
Add OMS-related models

### PR DESCRIPTION
**Pull request by eugenk**
This adds all the OMS-related models that are managed by Hets.

For many models, the tests only exist to test the factories that will certainly be used in the backend. Otherwise, associations are tested.

Please take a very close look at the migration as I have added `on_delete` clauses. Please check if they make sense. The [class diagram in the Wiki](https://github.com/ontohub/ontohub-backend/wiki/Architecture-Details#models) might help.

### CODE REVIEW

File: app/models/document.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+# The Document model contains OMS
+class Document < LocIdBase
+  one_to_many :document_links_source, dataset: (proc do |reflection|
```
Comment (phyrog) [inline app/models/document.rb]: Maybe `document_links_by_source` or `document_links_from_source`?

File: app/models/language.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+# The Language model
+class Language < Sequel::Model
+  one_to_many :logics
+  one_to_many :oms, class: OMS
+
+  one_to_many :language_mappings_source, dataset: (proc do |reflection|
```
Comment (phyrog) [inline app/models/language.rb]: See comment in `app/models/document.rb`.

File: app/models/language.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    reflection.associated_dataset.
+      where(Sequel[:language_mappings][:source_id] => id)
+  end), class: LanguageMapping
+
+  one_to_many :language_mappings_target, dataset: (proc do |reflection|
```
Comment (phyrog) [inline app/models/language.rb]: See comment in `app/models/document.rb`.

File: app/models/logic.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+# The Logic model
+class Logic < Sequel::Model
+  many_to_one :language
+  one_to_many :oms, class: OMS
+
+  one_to_many :logic_mappings_source, dataset: (proc do |reflection|
```
Comment (phyrog) [inline app/models/logic.rb]: See comment in `app/models/document.rb`.

File: app/models/logic.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    reflection.associated_dataset.
+      where(Sequel[:logic_mappings][:source_id] => id)
+  end), class: LogicMapping
+
+  one_to_many :logic_mappings_target, dataset: (proc do |reflection|
```
Comment (phyrog) [inline app/models/logic.rb]: See comment in `app/models/document.rb`.

File: app/models/oms.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  many_to_one :serialization
+  many_to_one :signature
+  many_to_one :normal_form, class: OMS
+  many_to_one :normal_form_signature_morphism, class: SignatureMorphism
+  many_to_one :free_normal_form, class: OMS
+  many_to_one :free_normal_form_signature_morphism, class: SignatureMorphism
+  many_to_one :cons_status
+  many_to_one :name_file_range, class: FileRange
+
+  one_to_many :mappings_source, dataset: (proc do |reflection|
```
Comment (phyrog) [inline app/models/oms.rb]: See comment in `app/models/document.rb`.

File: app/models/oms.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    reflection.associated_dataset.where(source_id: id)
+  end), class: Mapping
+
+  one_to_many :mappings_target, dataset: (proc do |reflection|
```
Comment (phyrog) [inline app/models/oms.rb]: See comment in `app/models/document.rb`.

File: app/models/signature.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            Sequel[:signature_symbols][:imported] => false)
+  end), class: OMSSymbol, join_table: :signature_symbols, right_key: :symbol_id
+
+  many_to_many :imported_symbols, dataset: (proc do |reflection|
+    reflection.associated_dataset.
+      where(Sequel[:signature_symbols][:signature_id] => id,
+            Sequel[:signature_symbols][:imported] => true)
+  end), class: OMSSymbol, join_table: :signature_symbols, right_key: :symbol_id
+
+  one_to_many :signature_morphisms_source, dataset: (proc do |reflection|
```
Comment (phyrog) [inline app/models/signature.rb]: See comment in `app/models/document.rb`.

File: app/models/signature.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    reflection.associated_dataset.
+      where(Sequel[:signature_morphisms][:source_id] => id)
+  end), class: SignatureMorphism
+
+  one_to_many :signature_morphisms_target, dataset: (proc do |reflection|
```
Comment (phyrog) [inline app/models/signature.rb]: See comment in `app/models/document.rb`.

File: spec/factories/cons_statuses_factory.rb [hunk 10]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+FactoryGirl.define do
+  factory :cons_status do
+    required do
+      %w(inconsistent string none pcons cons mono def)
```
Comment (phyrog) [inline spec/factories/cons_statuses_factory.rb]: Please use `sample`

File: spec/factories/cons_statuses_factory.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+    end
+    proved do
+      %w(inconsistent string none pcons cons mono def)
```
Comment (phyrog) [inline spec/factories/cons_statuses_factory.rb]: See above

File: spec/factories/diagnoses_factory.rb [hunk 12]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+FactoryGirl.define do
+  factory :diagnosis, class: Debug do
+    association :file_version
+    association :file_range
+    kind { [Error, Warn, Hint, Debug].map(&:to_s).sample }
```
Comment (phyrog) [inline spec/factories/diagnoses_factory.rb]: Nitpick: `[Error, Warn, Hint, Debug].sample.to_s` might be slightly faster :grin:

File: spec/factories/documents_factory.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+FactoryGirl.define do
+  factory :document do
+    association :file_version
+    kind { [Library, NativeDocument].map(&:to_s).sample }
```
Comment (phyrog) [inline spec/factories/documents_factory.rb]: Same nitpick as above

File: spec/factories/sentences_factory.rb [hunk 14]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+FactoryGirl.define do
+  factory :sentence, class: Axiom do
+    association :oms
+    association :file_range
+    kind { [Axiom, OpenConjecture, CounterTheorem, Theorem].map(&:to_s).sample }
```
Comment (phyrog) [inline spec/factories/sentences_factory.rb]: another nitpick

File: spec/factories/symbols_factory.rb [hunk 15]
```diff
@@ -0,0 +1,17 @@
+
+# frozen_string_literal: true
+
+FactoryGirl.define do
+  factory :symbol, class: OMSSymbol do
+    association :oms, factory: :oms
+    association :file_range
+    symbol_kind { %w(Op Pred Proposition Class Individual Name).sample }
```
Comment (phyrog) [inline spec/factories/symbols_factory.rb]: You're using `[..].map {}.sample` in other places. We should be consistent here

File: spec/models/logic_mapping_spec.rb [hunk 16]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe LanguageMapping do
```
Comment (phyrog) [inline spec/models/logic_mapping_spec.rb]: `LogicMapping`

File: spec/factories/file_versions_factory.rb [hunk 17]
```diff
@@ -5,5 +5,10 @@
     association :repository
     commit_sha { Faker::Crypto.sha1 }
     path { generate(:filepath) }
+
+    after(:create) do |file_version|
```
Comment (phyrog) [inline spec/factories/file_versions_factory.rb]: Can you add a todo comment that we want to run this asyncronously later on?

File: spec/factories/symbols_factory.rb (hunk 15 above)
Comment (eugenk) [reply spec/factories/symbols_factory.rb]: These are not subclasses (hence, the differing column name) ;)

File: spec/factories/file_versions_factory.rb (hunk 17 above)
Comment (eugenk) [reply spec/factories/file_versions_factory.rb]: That's for the backend to do ;)

**Comment by eugenk** [comment]
I just updated the class diagram

**Pull request merged**